packages/block-editor/src/components/use-on-block-drop/index.js

```diff
@@ -44,75 +45,83 @@ export function parseDropEvent( event ) {
 /**
  * A function that returns an event handler function for block drop events.
  *
- * @param {string}   targetRootClientId        The root client id where the block(s) will be inserted.
- * @param {number}   targetBlockIndex          The index where the block(s) will be inserted.
+ * @param {string} targetRootClientId        The root client id where the block(s) will be inserted.
+ * @param {number} targetBlockIndex          The index where the block(s) will be inserted.
  * @param {Function} getBlockIndex             A function that gets the index of a block.
  * @param {Function} getClientIdsOfDescendants A function that gets the client ids of descendant blocks.
  * @param {Function} moveBlocksToPosition      A function that moves blocks.
- *
+ * @param {Function} insertBlocks              A function that inserts blocks.
+ * @param {Function} clearSelectedBlock        A function that clears block selection.
  * @return {Function} The event handler for a block drop event.
  */
 export function onBlockDrop(
 	targetRootClientId,
 	targetBlockIndex,
 	getBlockIndex,
 	getClientIdsOfDescendants,
-	moveBlocksToPosition
+	moveBlocksToPosition,
+	insertBlocks,
+	clearSelectedBlock
 ) {
 	return ( event ) => {
 		const {
 			srcRootClientId: sourceRootClientId,
 			srcClientIds: sourceClientIds,
 			type: dropType,
+			blocks,
 		} = parseDropEvent( event );
 
-		// If the user isn't dropping a block, return early.
-		if ( dropType !== 'block' ) {
-			return;
+		// If the user is inserting a block
+		if ( dropType === 'inserter' ) {
+			clearSelectedBlock();
+			insertBlocks( blocks, targetBlockIndex, targetRootClientId, false );
 		}
 
-		const sourceBlockIndex = getBlockIndex(
-			sourceClientIds[ 0 ],
-			sourceRootClientId
-		);
-
-		// If the user is dropping to the same position, return early.
-		if (
-			sourceRootClientId === targetRootClientId &&
-			sourceBlockIndex === targetBlockIndex
-		) {
-			return;
-		}
+		// If the user is moving a block
+		if ( dropType === 'block' ) {
```

Comment:
Feels like it might be a good opportunity to rename these types. One thing is that it doesn't feel logical that the type `block` doesn't have the `blocks` property but `inserter` does.

What do you think about `moveBlocks` and `insertBlocks` to more closely reflect the actions?